### ISSUE TITLE
Worker fails silently and re-initializes

### ISSUE THREAD

**Issue by dannyshisler** [irothschild/stefuna#13]
I'm trying to run a task that does some audio and video processing, and I'm using the library librosa (https://librosa.github.io/librosa/) to do some audio analysis.

When I try to load the audio file with librosa, the worker ends silently and logs the message: 
```[INFO][ForkPoolWorker-5/20516] Initializing worker```

The code is:
```python
# Load the audio as a waveform `y`
# Store the sampling rate as `sr`
y, sr = librosa.load(audio_file)
```

Also, even thought it says ' Initializing worker', from that point onwards the server as stopped listening for new activities (until I stop and start the server).

The code and this library are working fine in all other environment (EC2/Docker etc), although the code is not threaded anywhere else.

**Comment by irothschild** [comment]
Try running with `--loglevel=debug` to see more detail. How many workers are you running? How do you know the worker ends? 

You will only be able to process new activities when there is a worker available. Each worker is a separate process so threading shouldn't be a problem.

**Comment by dannyshisler** [comment]
Hi, I've tried running it with ```--loglevel=debug``` and it doesn't give me any more information, when it gets to the line that tries to load the file in librosa, nothing happens, execution ends and there is a log statement that says ```[INFO][ForkPoolWorker-2/23006] Initializing worker```.

I'm only running 1 worker at the moment.

**Comment by irothschild** [comment]
If an exception is thrown during the Worker processing a task it *should* be caught automatically (and the activity task will be be failed with `Task.Failure`. I've run workers with various libraries with no problems. I will do a quick try with librosa and get back to you.

**Comment by dannyshisler** [comment]
Ok, a little more info - I'm developing on a Mac. The process fails when it tries to use the macOS CoreAudio library to load the audio file.

It fails on this line in the AudioRead library that librosa depends on:
```python
check(_coreaudio.ExtAudioFileOpenURL(
            url._obj, ctypes.byref(file_obj)
        ))
```
https://github.com/beetbox/audioread/blob/master/audioread/macca.py#L219

I think there are 2 issues here: why does this call fail when run in stefuna (execution as part of a thread pool is the only difference I can see between this execution context and what's been working for me previously)? And why isn't the exception caught and handled/reported when it does occur?

**Comment by irothschild** [comment]
I can reproduce the process dying when the `audioread.audio_open()` is called on OSX. The code does not seem to be raising an exception.  I think it's caused by codeaudio being called from a child process (which the worker is) and it seems it may have to do with not calling `exec` on the child process. I remember there was a potential issue with this and the python worker pool. I will have to do more research and see what the current status of this is.

**Comment by dannyshisler** [comment]
OK, thanks for your help, much appreciated.

**Comment by irothschild** [comment]
Here is a fix. I need to do some testing before I make the change to the code and I may do it as a config option.

Fork this repo. In `stefuna/stefuna.py` change line 7 to be:

```
from multiprocessing import cpu_count, set_start_method  # noqa
```
And add a line at line 40 (right after `def main():`):

```
    set_start_method('spawn')
```

**Comment by dannyshisler** [comment]
Fantastic, I'll give it a go. Thanks!

**Comment by irothschild** [comment]
Here's more background on the issue with OSX https://bugs.python.org/issue33725

**Comment by dannyshisler** [comment]
That works - it completely fixes the issue. 

The background is interesting - I've only had a chance to have a quick look, but is my understanding correct that Python Pools Fork by default rather than Spawn which can lead to unexpected errors if the child process calls Objective-C code?

Thanks very much again for your help.

**Comment by irothschild** [comment]
Fixed in https://github.com/irothschild/stefuna/pull/14

**Comment by irothschild** [comment]
@dannyshisler Just using `set_start_method('spawn')` fixes the problem but it introduces a worker initialization issue. This is now fixed in 0.9.9. Latest release is 1.0.0 which you can install with pip.

**Comment by dannyshisler** [comment]
Thanks a lot @irothschild I'll give it a go
